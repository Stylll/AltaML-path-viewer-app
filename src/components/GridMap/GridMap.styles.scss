.gridContainer {
    display: flex;
    justify-content: center;
    align-items: center;
    min-height: 50vh;
    margin-bottom: 50px;

    .gridTitle {
        font-size: 24px;
    }
}

.legendContainer {
    &__legendRow {
        display: flex;
        justify-content: center;
        &--col {
            display: flex;
            margin: 10px;
            align-items: center;
            .td {
                margin-right: 5px;
            }
        }
    }

    &__legendTitle {
        font-size: 18px;
    }
}

td {
    height: 30px;
    width: 50px;
    background-color: #cff3e7;
}

.td {
    height: 50px;
    width: 50px;
}

.active {
    background-color: rgb(16, 177, 129);
}

.multi {
    background-color: #114B5F;
}

.start {
    background-repeat: no-repeat;
    background-size: cover;
    background-image: linear-gradient(rgba(37, 42, 48, 0.6), rgba(37, 42, 48, 0.6)),
    url("../../assets/flag-variant.png");
}

.stop {
    background-color: #89043D;
    background-repeat: no-repeat;
    background-size: cover;
    background-image: linear-gradient(rgba(37, 42, 48, 0.2), rgba(37, 42, 48, 0.2)),
    url("../../assets/turtle.png");
}

.left {
    transform: rotate(-90deg);
}

.right {
    transform: rotate(90deg);
}

.down {
    transform: rotate(180deg);
}

.loading {
    background: url("../../assets/loading.gif");
    background-repeat: no-repeat;
    background-size: cover;
    height: 50px;
    width: 50px;
}

.errorContainer {
    background-color: #D11919;
    color: #F9F8F8;
    padding: 20px;
    p {
        margin: 0;
    }
}
