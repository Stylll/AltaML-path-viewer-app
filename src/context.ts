import { createContext, Context } from 'react';

import { IAppContext } from './types';

export const AppContext: Context<IAppContext>  = createContext<IAppContext>({
    direction: undefined,
});
