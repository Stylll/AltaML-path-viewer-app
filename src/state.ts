import { IDirection, IState } from './types';

export const direction: IDirection = {
    bound: {
        posXBound: [-2,-1,0,1,2,3],
        posYBound: [1,0,-1],
    },
    endPosition: {
        raw: [3,1],
        string: '3/1',
    },
    highestPosition: { x: 3,
        y: 1 },
    lastDirection: 'up',
    lowestPosition: { x: -2,
        y: -1 },
    multiOccurence: ['-2/-1'],
    path: {
        raw: [
            [0,0],
            [0,1],
            [-1,1],
            [-2,1],
            [-2,0],
            [-2,-1],
            [-1,-1],
            [0,-1],
            [1,-1],
            [1,0],
            [2,0],
            [3,0],
            [3,1],
        ],
        string: ['0/0','0/1','-1/1','-2/1','-2/0','-2/-1','-1/-1','0/-1','1/-1','1/0','2/0','3/0','3/1'],
    },
    startPosition: {
        raw: [0,0],
        string: '0/0',
    },
};

const directionNames: string[] = [
    'directions-1',
    'directions-2',
    'directions-3',
    'directions-4',
    'directions-5',
    'directions-6',
];

export const defaultState: IState = {
    direction: undefined,
    directionNames: directionNames,
};
