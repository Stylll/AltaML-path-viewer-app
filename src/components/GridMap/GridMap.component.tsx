import React, { ReactElement, useContext } from 'react';

import './GridMap.styles.scss';
import { AppContext } from '../../context';
import { IAppContext, IDirection } from '../../types';

export function GridMap (): ReactElement {
    const { direction, isLoading, error }: IAppContext = useContext(AppContext);

    function generateCellClass (posXY: string, direction: IDirection): string {

        let cellClass: string = '';

        const multi: string | undefined = direction.multiOccurence.find((value: string): boolean => {
            return value === posXY;
        });

        const start: boolean = direction.startPosition.string === posXY;
        const stop: boolean = direction.endPosition.string === posXY;

        if (multi) {
            cellClass = 'multi';
        }

        if (start) {
            cellClass = `${cellClass} start`;
        } else if (stop) {
            cellClass = `${cellClass} stop ${direction.lastDirection}`;
        }

        if (multi || start || stop) {
            return cellClass;
        }

        const isValidPath: string | undefined = direction.path.string.find((value: string): boolean => {
            return value === posXY;
        });

        if (isValidPath) {
            cellClass = `${cellClass} active`;
        }

        return cellClass;
    }

    function renderCell (posXY: string, direction: IDirection): ReactElement {
        const cellClass: string = generateCellClass(posXY, direction);

        return (
            <td key={posXY} className={cellClass}></td>
        );
    }

    function renderGrid (direction: IDirection): ReactElement {
        return (
            <table>
                <tbody>
                    {direction && direction.bound.posYBound.map((posY: number): ReactElement => {
                        const y: string = `${posY}`;

                        return (
                            <tr key={posY}>
                                {direction.bound.posXBound.map((posX: number): ReactElement => {
                                    const x: string = `${posX}`;
                                    const xy: string = `${x}/${y}`;

                                    return renderCell(xy, direction);
                                })}
                            </tr>
                        );
                    })}
                </tbody>
            </table>
        );
    }

    function RenderLegend (): ReactElement {
        return (
            <div className="legendContainer">
                <p className="legendContainer__legendTitle">Legend</p>
                <div className="legendContainer__legendRow">
                    <div className="legendContainer__legendRow--col">
                        <div className="td active"></div>
                        <span>Travel Path</span>
                    </div>
                    <div className="legendContainer__legendRow--col">
                        <div className="td multi"></div>
                        <span>Cross Path</span>
                    </div>
                    <div className="legendContainer__legendRow--col">
                        <div className="td start"></div>
                        <span>Start Position</span>
                    </div>
                    <div className="legendContainer__legendRow--col">
                        <div className="td stop"></div>
                        <span>End Position</span>
                    </div>
                </div>
            </div>
        );
    }

    function renderError (error: string): ReactElement {
        return (
            <div className="errorContainer">
                <p>{error}</p>
            </div>
        );
    }

    function RenderIsLoading (): ReactElement {
        return (
            <div className="loading" />
        );
    }

    return (
        <div>
            {error && renderError(error)}
            <RenderLegend />
            <br/>
            <div className="gridContainer">
                {isLoading && <RenderIsLoading />}
                {!isLoading && <div>
                    <p className="gridTitle">Path</p>
                    {direction && renderGrid(direction)}
                    {!direction && <p>No Path Available</p>}
                </div>}
            </div>
        </div>
    );
}
