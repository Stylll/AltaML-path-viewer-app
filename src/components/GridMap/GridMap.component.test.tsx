import { render } from '@testing-library/react';
import React from 'react';

import { AppContext } from '../../context';
import { defaultState, direction } from '../../state';

import { GridMap } from './GridMap.component';

describe('GridMap Component', (): void => {

    it('should render properly', (): void => {
        const { baseElement } = render(
            <AppContext.Provider value={
                {
                    direction: direction,
                    loadDirection: jest.fn(),
                    directionNames: defaultState.directionNames,
                    isLoading: false,
                    error: '',
                }
            }
            >
                <GridMap />
            </AppContext.Provider>
        );

        expect(baseElement).toMatchSnapshot();
    });

    it('should display error message when there\'s an error ', (): void => {
        const { getByText } = render(
            <AppContext.Provider value={
                {
                    direction: direction,
                    loadDirection: jest.fn(),
                    directionNames: defaultState.directionNames,
                    isLoading: false,
                    error: 'Error exists',
                }
            }
            >
                <GridMap />
            </AppContext.Provider>
        );

        expect(getByText('Error exists')).toBeInTheDocument();
    });

    it('should only display loader when loading ', (): void => {
        const { container, queryByText } = render(
            <AppContext.Provider value={
                {
                    direction: direction,
                    loadDirection: jest.fn(),
                    directionNames: defaultState.directionNames,
                    isLoading: true,
                    error: '',
                }
            }
            >
                <GridMap />
            </AppContext.Provider>
        );

        expect(queryByText('Path')).not.toBeInTheDocument();
        expect(queryByText('No Path Available')).not.toBeInTheDocument();
        expect(container.querySelector('table')).not.toBeInTheDocument();
    });
});
