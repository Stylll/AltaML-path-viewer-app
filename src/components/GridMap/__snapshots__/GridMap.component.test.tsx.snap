// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridMap Component should render properly 1`] = `
<body>
  <div>
    <div>
      
      <div
        class="legendContainer"
      >
        <p
          class="legendContainer__legendTitle"
        >
          Legend
        </p>
        <div
          class="legendContainer__legendRow"
        >
          <div
            class="legendContainer__legendRow--col"
          >
            <div
              class="td active"
            />
            <span>
              Travel Path
            </span>
          </div>
          <div
            class="legendContainer__legendRow--col"
          >
            <div
              class="td multi"
            />
            <span>
              Cross Path
            </span>
          </div>
          <div
            class="legendContainer__legendRow--col"
          >
            <div
              class="td start"
            />
            <span>
              Start Position
            </span>
          </div>
          <div
            class="legendContainer__legendRow--col"
          >
            <div
              class="td stop"
            />
            <span>
              End Position
            </span>
          </div>
        </div>
      </div>
      <br />
      <div
        class="gridContainer"
      >
        <div>
          <p
            class="gridTitle"
          >
            Path
          </p>
          <table>
            <tbody>
              <tr>
                <td
                  class=" active"
                />
                <td
                  class=" active"
                />
                <td
                  class=" active"
                />
                <td
                  class=""
                />
                <td
                  class=""
                />
                <td
                  class=" stop up"
                />
              </tr>
              <tr>
                <td
                  class=" active"
                />
                <td
                  class=""
                />
                <td
                  class=" start"
                />
                <td
                  class=" active"
                />
                <td
                  class=" active"
                />
                <td
                  class=" active"
                />
              </tr>
              <tr>
                <td
                  class="multi"
                />
                <td
                  class=" active"
                />
                <td
                  class=" active"
                />
                <td
                  class=" active"
                />
                <td
                  class=""
                />
                <td
                  class=""
                />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;
